### PR TITLE
[FEA] Exposing python wrapper for the RMAT generator logic

### PR DESCRIPTION
**Pull request by teju85**
This PR adds a thin python wrapper for the RMAT generator logic (as per the request from @PiotrBigajNV)

@cjnolet I was not sure what's the right way to wrap other such RAFT prims to python layer. Hence, I just followed the same path as that of `raft_distance` and created a `raft_random` build target instead. Let me know if you have a better approach for this one.

### CODE REVIEW

File: python/pylibraft/pylibraft/test/test_random.py
Comment (cjnolet) [inline python/pylibraft/pylibraft/test/test_random.py:24]: Since we're using this in two places now, we should probably centralize it in its own file so we can update it in one place when needed.

**Review by cjnolet — CHANGES_REQUESTED** [review]
Super excited to start exposing more of our primitives through Python! Other than a discussion on the shared libs / components, I only have minor feedback

File: build.sh
Comment (teju85) [reply build.sh:200]: This was the exact thing I wasn't sure of! I'm glad to know that you'd like to keep this generic. Let me propose a change.

File: python/pylibraft/pylibraft/test/test_random.py
Comment (teju85) [reply python/pylibraft/pylibraft/test/test_random.py:24]: Done

File: python/pylibraft/pylibraft/test/test_random.py
Comment (teju85) [reply python/pylibraft/pylibraft/test/test_random.py:50]: Good point. Done.

File: python/pylibraft/pylibraft/random/rmat_rectangular_generator.pyx
Comment (teju85) [reply python/pylibraft/pylibraft/random/rmat_rectangular_generator.pyx:84]: Oops.. that was an oversight. I was doing this when I was trying to support generating both AoS and SoA output formats with a single call (c++ API supports this). But then decided to settle to AoS format for the python layer, to begin with. Thanks for catching this. Fixed the docstring.

File: build.sh
Comment (teju85) [reply build.sh]: Can you check now @cjnolet ?

Comment (cjnolet) [reply build.sh]: I like it a lot- so users have the option of header-only device or runtime APIs or shared lib w/ host accessable API

**Comment by cjnolet** [comment]
Not something we need to do for this PR, but we might also consider having a separate section in the raft docs for the `header-only` APIs and the `host accessable APIs`.

**Review by cjnolet — APPROVED** [review]

**Comment by teju85** [comment]
@cjnolet and/or @vyasr can one of you help me dealing with the CI errors please? Unable to figure out what I'm doing wrong here. :(

File: cpp/include/raft_compiled/random/rmat_rectangular_generator.hpp
Comment (lowener) [inline cpp/include/raft_compiled/random/rmat_rectangular_generator.hpp:19]:
```suggestion
#include <raft/core/handle.hpp>
```

**Comment by teju85** [comment]
@cjnolet and @vyasr This PR doesn't touch the "distance-lib" but the following errors can be seen in the CI and I'm not sure why this is happening. Any suggestions/ideas?
```
CMake Error at /workspace/.conda-bld/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/cmake/raft/raft-distance-lib-targets.cmake:61 (set_target_properties):
  The link interface of target "raft::raft_distance_lib" contains:

    cuco::cuco

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.
```

**Comment by cjnolet** [comment]
@vyasr do you notice anything obvious or stange in the build? What's weird is that it seems to be looking for that target and I'd assume that would also have caused it to fail when running the `find_package` for cuco, but it doesn't seem to find that.

What's even more strange is that it only seems to be happening in CI. 

```
if(RAFT_COMPILE_DIST_LIBRARY OR distance IN_LIST raft_FIND_COMPONENTS)
  set(RAFT_ENABLE_cuco_DEPENDENCY ON)
endif()
```

**Comment by teju85** [comment]
@vyasr and/or @robertmaynard when you get time, can one of you guys help us here? I'm not sure why CI is failing here!

**Comment by cjnolet** [comment]
@teju85 I'm thinking maybe for now we should just have `libraft-distance.so` compile the `random` source files as well and then add a note/todo in the cmakelists that we're eventually going to be renaming `libraft-distance.so` to `libraft.so` (and likely merging in the nearest neighbors source files as well once we remove faiss).

File: cpp/CMakeLists.txt
Comment (robertmaynard) [inline cpp/CMakeLists.txt:479]:
```suggestion
set(raft_install_comp raft raft raft)
```

The `raft_components` and `raft_install_comp` need to be equal lengths

Comment (robertmaynard) [reply cpp/CMakeLists.txt:479]: I expect this is the bug as it would cause the final sets to look like:

```
-- raft-distance-targets.cmake => install component raft
-- raft-nn-targets.cmake => install component raft
-- raft-compiled-targets.cmake => install component distance
-- raft-distance-lib-targets.cmake => install component nn
-- raft-nn-lib-targets.cmake => install component compiled
-- raft-compiled-lib-targets.cmake => install component 
```

**Review by robertmaynard — CHANGES_REQUESTED** [review]

File: cpp/CMakeLists.txt
Comment (teju85) [reply cpp/CMakeLists.txt]: Duh... such a silly mistake. Sorry about that. Thanks a lot @robertmaynard ! This has fixed the build issue for me. Now somehow the pytest isn't able to find the right modules. I'll figure out that.

**Comment by cjnolet** [comment]
@teju85 Looks like we got it finally! I think I traced the issue to `__init__.py` files that shouldn't have been created (one in the root of `pylibraft` and another in the `tests` directory. I think we're good to go at this point, pending @robertmaynard's blessing, of course.

**Comment by teju85** [comment]
woohoo! Thanks @cjnolet and @robertmaynard . We need a final approval from @robertmaynard before this can be merged.

**Review by robertmaynard — APPROVED** [review]

**Comment by cjnolet** [comment]
@gpucibot merge

**Review by cjnolet — APPROVED** [review]

**Review by Ethyling — APPROVED** [review]

**Pull request merged**
